src/overlays/actors/ovl_En_Ex_Ruppy/z_en_ex_ruppy.c

```diff
@@ -21,29 +34,345 @@ const ActorInit En_Ex_Ruppy_InitVars = {
+        //Normal rupee
+         osSyncPrintf(VT_FGCOL(GREEN) "☆☆☆☆☆ ノーマルルピー ☆☆☆☆☆ \n" VT_RST);
+        this->actor.shape.unk_10 = 6.0f;
+        this->actor.shape.unk_08 = 700.0f;
+        this->actor.flags &= ~1;
+        this->actionFunc = EnExRuppy_Collect;
+        
+        break;
+    case 4:
+        this->actor.gravity = -3.0f;
+        this->actor.flags &= ~1;
+        Actor_SetScale(thisx, 0.01f);
+        this->actor.shape.unk_10 = 6.0f;
+        this->actor.shape.unk_08 = -700.0f;
+        this->actionFunc = func_80A0B0F4;
+        break;
+    }
+}
+#else
 #pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Ex_Ruppy/EnExRuppy_Init.s")
+#endif
+
+void EnExRuppy_Destroy(Actor *thisx, GlobalContext *globalCtx) {
+}
+
+void EnExRuppy_SpawnSparkles(EnExRuppy *this, GlobalContext *globalCtx, s16 arg2, s32 arg3) {//
+    Vec3f tempPos;
+    Vec3f tempVelocity;
+    Vec3f tempAccel;
+    Color_RGBA8_n tempPrimColor;
+    Color_RGBA8_n tempEnvColor;
+    s32 i;
+    s16 tempScale;
+    s16 tempLife;
+
+    if (arg2 <= 0) {
+        arg2 = 1;
+    }
+    tempPrimColor.r = 255;
+    tempPrimColor.g = 255;
+    tempPrimColor.b = 0;
+    tempEnvColor.r = 255;
+    tempEnvColor.g = 255;
+    tempEnvColor.b = 255;
+    tempVelocity = D_80A0B358[arg3];
+    tempAccel = D_80A0B370[arg3];
+    tempScale = 3000;
+    tempLife = 16;
+    for (i= 0 ; i < arg2 ; i++){
+        if (arg3 == 1) {
+            tempAccel.x = Math_Rand_CenteredFloat(20.0f);
+            tempAccel.z = Math_Rand_CenteredFloat(20.0f);
+            tempScale = 5000;
+            tempLife = 20;
+        }
+        tempPos.x = ((Math_Rand_ZeroOne() - 0.5f) * 10.0f) + this->actor.posRot.pos.x;
+        tempPos.y = ((Math_Rand_ZeroOne() - 0.5f) * 10.0f) + (this->actor.posRot.pos.y + (this->unk_160 * 600.0f));
+        tempPos.z = ((Math_Rand_ZeroOne() - 0.5f) * 10.0f) + this->actor.posRot.pos.z;
+        func_80028BB0(globalCtx, &tempPos, &tempVelocity, &tempAccel, &tempPrimColor, &tempEnvColor, tempScale, tempLife);
+    }
+}
+
+void EnExRuppy_DropIntoWater(EnExRuppy *this, GlobalContext *globalCtx) {
+    this->actor.shape.rot.y = (this->actor.shape.rot.y + 1960);
+    Math_SmoothScaleMaxF(&this->actor.gravity, -2.0f, 0.3f, 1.0f);
+    EnExRuppy_SpawnSparkles(this, globalCtx, 2, 0);
+    func_80078884(NA_SE_EV_RAINBOW_SHOWER - SFX_FLAG);
+    if (((DIVING_GAME != NULL) && (DIVING_GAME->actor.update != NULL)) && (((DIVING_GAME->unk_296 == 0) || (this->actor.bgCheckFlags & 0x20)) || (this->timer == 0))) {
+        this->isFalling = 1;
+        this->actor.speedXZ = 0.0f;
+        this->actor.velocity.z = 0.0f;
+        this->actor.velocity.y = 0.0f;
+        this->actor.velocity.x = 0.0f;
+        this->actor.gravity = 0.0f;
+        func_80078914(&this->actor.projectedPos, NA_SE_EV_BOMB_DROP_WATER);
+        this->actionFunc = EnExRuppy_ChoseFinalPos;
+    } 
+}
+
+void EnExRuppy_ChoseFinalPos(EnExRuppy *this, GlobalContext *globalCtx) {
+    s32 pad;
```

Comment:
Im on the fence about this being a good action name, but I cant think of anything else at the moment. 
I do think you meant Choose and instead of Chose tho